src/gt4py/next/iterator/type_system/inference.py

```diff
@@ -0,0 +1,606 @@
+    ... )
+    True
+
+    Note that this is a simple example where the type of the arguments and the return value is
+    available when the function is called. In order to support higher-order functions, where
+    arguments or return value are functions itself (i.e. passed as type rules) this class provides
+    additional functionality for multiple typing rules to notify each other about a type being
+    ready.
+    """
+
+    #: type rule that given a set of types or type rules returns the return type or a type rule
+    type_rule: rules.TypeInferenceRule
+    #: offset provider used by some type rules
+    offset_provider: common.OffsetProvider
+    #: node that has this type
+    node: Optional[itir.Node] = None
+    #: list of references to this function
+    aliases: list[itir.SymRef] = dataclasses.field(default_factory=list)
+    #: list of callbacks executed as soon as the type is ready
+    callbacks: list[Callable[[ts.TypeSpec], None]] = dataclasses.field(default_factory=list)
+    #: the inferred type when ready and None until then
+    inferred_type: Optional[ts.FunctionType] = None
+    #: whether to store the type in the node or not
+    store_inferred_type_in_node: bool = False
+
+    def infer_type(
+        self, return_type: ts.DataType | ts.DeferredType, *args: ts.DataType | ts.DeferredType
+    ) -> ts.FunctionType:
+        return ts.FunctionType(
+            pos_only_args=list(args), pos_or_kw_args={}, kw_only_args={}, returns=return_type
+        )
+
+    def _infer_type_listener(self, return_type: ts.TypeSpec, *args: ts.TypeSpec) -> None:
+        self.inferred_type = self.infer_type(return_type, *args)  # type: ignore[arg-type]  # ensured by assert above
+
+        # if the type has been fully inferred, notify all `ObservableTypeInferenceRule`s that depend on it.
+        for cb in self.callbacks:
+            cb(self.inferred_type)
+
+        if self.store_inferred_type_in_node:
+            assert self.node
+            _set_node_type(self.node, self.inferred_type)
+            self.node.type = self.inferred_type
+            for alias in self.aliases:
+                _set_node_type(alias, self.inferred_type)
+
+    def on_type_ready(self, cb: Callable[[ts.TypeSpec], None]) -> None:
+        if self.inferred_type:
+            # type has already been inferred, just call the callback
+            cb(self.inferred_type)
+        else:
+            self.callbacks.append(cb)
+
+    def __call__(
+        self, *args: Union[ts.TypeSpec, "ObservableTypeInferenceRule"]
+    ) -> Union[ts.TypeSpec, "ObservableTypeInferenceRule"]:
+        if "offset_provider" in inspect.signature(self.type_rule).parameters:
+            return_type = self.type_rule(*args, offset_provider=self.offset_provider)
+        else:
+            return_type = self.type_rule(*args)
+
+        # return type is a typing rule by itself
+        if callable(return_type):
```

Comment:
Just a comment: checking only if it's callable is slightly dangerous because it's a bit too general and returning by mistake things like a type would be interpreted as a type rule because types are callables (the constructor)